File: design-pattern/flyweight-pattern/flyweight.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <unordered_map>
#include <random>
#include <string>
#include <chrono>

using namespace std;

class AsteriodFlyweight
{
private:
    int length;
    int width;
    int weight;
    string color;
    string texture;
    string material;

public:
    AsteriodFlyweight(int l, int w, int wt, string c, string t, string m)
        : length(l), width(w), weight(wt), color(c), texture(t), material(m) {}

    void render(int posX, int posY, int velocityX, int velocityY) const
    {
        cout << "Rendering Asteriod at (" << posX << ", " << posY << ") with velocity ("
             << velocityX << ", " << velocityY << ") and properties: "
             << "Length: " << length << ", Width: " << width
             << ", Weight: " << weight << ", Color: " << color
             << ", Texture: " << texture << ", Material: " << material
             << endl;
    }

    static size_t getMemoryUsage()
    {
        return sizeof(int) * 3 + sizeof(string) * 3 + 32 * 3;
    }
};

class AsteriodFactory
{
private:
    static unordered_map<string, AsteriodFlyweight *> asteriods;

public:
    static AsteriodFlyweight *getAsteriod(int length, int width, int weight, string color, string texture, string material)
    {
        string key = to_string(length) + "-" + to_string(width) + "-" + to_string(weight) + "-" + color + "-" + texture + "-" + material;

        if (asteriods.find(key) == asteriods.end())
        {
            asteriods[key] = new AsteriodFlyweight(length, width, weight, color, texture, material);
        }

        return asteriods[key];
    }

    static int getFlyweightCount()
    {
        return asteriods.size();
    };

    static size_t getTotalMemoryUsage()
    {
        return asteriods.size() * AsteriodFlyweight::getMemoryUsage();
    };

    static void cleanup()
    {
        asteriods.clear();
    }
};

unordered_map<string, AsteriodFlyweight *> AsteriodFactory::asteriods;

class AsteriodContext
{
private:
    int posX, posY;
    int velocityX, velocityY;
    AsteriodFlyweight *flyweight;

public:
    AsteriodContext(int x, int y, int vx, int vy, AsteriodFlyweight *fw)
        : posX(x), posY(y), velocityX(vx), velocityY(vy), flyweight(fw) {};

    void render()
    {
        flyweight->render(posX, posY, velocityX, velocityY);
    };

    static size_t getMemoryUsage()
    {
        return sizeof(int) * 4 + sizeof(AsteriodFlyweight *);
    };
};

class spaceGamewithFlyweight
{
private:
    static vector<AsteriodContext *> asteriods;

public:
    void spawnAsteriods(int count)
    {
        vector<string> color = {"Red", "Blue", "Gray"};
        vector<string> texture = {"Rocky", "Metallic", "Icy"};
        vector<string> material = {"Iron", "Silicon", "Carbon"};
        int size[] = {10, 20, 30};
        for (int i = 0; i < count; i++)
        {
            int type = i % 3;
            AsteriodFlyweight *flyweight = AsteriodFactory::getAsteriod(
                size[type], size[type], size[type] * 10,
                color[type], texture[type], material[type]);

            asteriods.push_back(new AsteriodContext(
                rand() % 100, rand() % 100, 1, 2, flyweight));
        }
    };

    void renderALL()
    {
        for (int i = 0; i < min(5, (int)asteriods.size()); i++)
        {
            asteriods[i]->render();
        }
    };

    static size_t calculateMemoryUsage()
    {
        size_t contextMemory = asteriods.size() * AsteriodContext::getMemoryUsage();
        size_t flyweightMemory = AsteriodFactory::getTotalMemoryUsage();
        return contextMemory + flyweightMemory;
    };

    int getAsteriodCount()
    {
        return asteriods.size();
    };
};

vector<AsteriodContext *> spaceGamewithFlyweight::asteriods;

int main(void)
{
    const int ASTRIOD_COUNT = 10000;

    cout << "Testing with Flyweight" << endl;
    spaceGamewithFlyweight *game = new spaceGamewithFlyweight();

    game->spawnAsteriods(ASTRIOD_COUNT);
    game->renderALL();

    size_t totalMemoryUsage = game->calculateMemoryUsage();

    cout << "\n === Memory Usage ===" << endl;
    cout << "Total Asteriods: " << ASTRIOD_COUNT << endl;

    cout << "Memory per Asteriod: " << AsteriodFlyweight::getMemoryUsage() << " bytes" << endl;

    cout << "Total Memory Usage: " << totalMemoryUsage << " bytes" << endl;
    cout << "Total Memory Usage: " << totalMemoryUsage / (1024) << "KM" << endl;

    AsteriodFactory::cleanup();
    return 0;
}

File: design-pattern/flyweight-pattern/flyweight.ts
class AsteriodFlyweight {
    private length: number;
    private width: number;
    private weight: number;
    private color: string;
    private texture: string;
    private material: string;

    constructor(l: number, w: number, wt: number, c: string, t: string, m: string) {
        this.length = l;
        this.width = w;
        this.weight = wt;
        this.color = c;
        this.texture = t;
        this.material = m;
    };

    render(posX: number, posY: number, velocityX: number, velocityY: number): void {
        console.log(
            `Rendering Asteriod at (${posX}, ${posY}) with velocity (${velocityX}, ${velocityY}) 
            Properties -> Length: ${this.length}, Width: ${this.width}, Weight: ${this.weight}, 
            Color: ${this.color}, Texture: ${this.texture}, Material: ${this.material}`
        );
    }


    static getMemoryUsage() {
        return 4 * 3 + 50 * 3;
    }
};

class AsteriodFactory {
    private static asteriods: Map<string, AsteriodFlyweight> = new Map();

    static getAsteriod(
        length: number,
        width: number,
        weight: number,
        color: string,
        texture: string,
        material: string
    ): AsteriodFlyweight {
        const key = `${length}-${width}-${weight}-${color}-${texture}-${material}`;

        if (!this.asteriods.has(key)) {
            this.asteriods.set(key, new AsteriodFlyweight(length, width, weight, color, texture, material));
        };

        return this.asteriods.get(key)!;
    };

    static getFlyweightCount(): number {
        return this.asteriods.size;
    };

    static getTotalMemoryUsage(): number {
        return this.asteriods.size * AsteriodFlyweight.getMemoryUsage();
    };



    static cleanup(): void {
        this.asteriods.clear();
    }
};

class AsteriodContext {
    private posX: number;
    private posY: number;
    private velocityX: number;
    private velocityY: number;
    private flyweight: AsteriodFlyweight;


    constructor(x: number, y: number, vx: number, vy: number, fw: AsteriodFlyweight) {
        this.posX = x;
        this.posY = y;
        this.velocityX = vx;
        this.velocityY = vy;
        this.flyweight = fw;
    };

    render(): void {
        this.flyweight.render(this.posX, this.posY, this.velocityX, this.velocityY)
    };

    static getMemoryUsage(): number {
        return 4 * 4 + 8; // approx
    }
};

class spaceGameWithFlyweight {
    private static asteriods: AsteriodContext[] = [];

    spawnAsteriods(count: number): void {
        const color = ["Red", "Blue", "Gray"];
        const texture = ["Rocky", "Metallic", "Icy"];
        const material = ["Iron", "Silicon", "Carbon"];
        const size = [10, 20, 30];

        for (let i = 0; i < count; i++) {
            const type = i % 3;

            const flyweight = AsteriodFactory.getAsteriod(
                size[type],
                size[type],
                size[type] * 10,
                color[type],
                texture[type],
                material[type]
            );

            spaceGameWithFlyweight.asteriods.push(
                new AsteriodContext(
                    Math.floor(Math.random() * 100),
                    Math.floor(Math.random() * 100),
                    1,
                    2,
                    flyweight
                )
            );
        }
    }

    renderAll(): void {
        for (let i = 0; i < Math.min(5, spaceGameWithFlyweight.asteriods.length); i++) {
            spaceGameWithFlyweight.asteriods[i].render();
        }
    };

    static calculateMemoryUsage(): number {
        const contextMemory =
            spaceGameWithFlyweight.asteriods.length *
            AsteriodContext.getMemoryUsage();

        const flyweightMemory = AsteriodFactory.getTotalMemoryUsage();

        return contextMemory + flyweightMemory;
    }

    getAsteriodCount(): number {
        return spaceGameWithFlyweight.asteriods.length;
    }
}

const ASTEROID_COUNT = 10000;

console.log("Testing with Flyweight");

const game = new spaceGameWithFlyweight();

game.spawnAsteriods(ASTEROID_COUNT);
game.renderAll();

const totalMemoryUsage = spaceGameWithFlyweight.calculateMemoryUsage();

console.log("\n=== Memory Usage ===");
console.log("Total Asteroids:", ASTEROID_COUNT);
console.log("Memory per Asteroid:", AsteriodFlyweight.getMemoryUsage(), "bytes");
console.log("Total Memory Usage:", totalMemoryUsage, "bytes");
console.log("Total Memory Usage:", totalMemoryUsage / 1024, "KB");

AsteriodFactory.cleanup();